pkg/sources/s3/s3.go

```diff
@@ -196,32 +213,91 @@ func (s *Source) getBucketsToScan(client *s3.S3) ([]string, error) {
 			bucketsToScan = append(bucketsToScan, name)
 		}
 	}
+	slices.Sort(bucketsToScan)
+
 	return bucketsToScan, nil
 }
 
+// workerSignal provides thread-safe tracking of cancellation state across multiple
+// goroutines processing S3 bucket pages. It ensures graceful shutdown when the context
+// is cancelled during bucket scanning operations.
+//
+// This type serves several key purposes:
+//  1. AWS ListObjectsV2PagesWithContext requires a callback that can only return bool,
+//     not error. workerSignal bridges this gap by providing a way to communicate
+//     cancellation back to the caller.
```

Comment:
I’ve been unable to get this function to return an error when the context is canceled. In my testing, the workers correctly handle the cancellation and stop, but this function itself doesn’t return an error. Because of this, I can’t propagate the error to the caller.

If my understanding is correct, when scanning a single bucket with many pages, this function is only called once. The s.pageChunker function gets invoked multiple times instead, meaning canceling the context for this function has no effect.

Maybe I’m missing something. Feel free to try it yourself to see if you get different results—I’ve probably hit my mental capacity for now. 😅

Reply:
Looking at the function signature, it seems the context isn’t propagated to the passed-in function.

```
// ListObjectsV2PagesWithContext same as ListObjectsV2Pages except
// it takes a Context and allows setting request options on the pages.
//
// The context must be non-nil and will be used for request cancellation. If
// the context is nil a panic will occur. In the future the SDK may create
// sub-contexts for http.Requests. See https://golang.org/pkg/context/
// for more information on using Contexts.
func (c *S3) ListObjectsV2PagesWithContext(ctx aws.Context, input *ListObjectsV2Input, fn func(*ListObjectsV2Output, bool) bool, opts ...request.Option) error {
....
}

Reply:
Would doing something like this work?

```go
err := ListObjectsV2PagesWithContext(ctx, ...)
if err == nil {
    err = ctx.Err()
}
```

Reply:
Removed ctx cancellation logic as per the Slack convo.